pkg/sql/colexec2/aggregate/aggregate.go

```diff
@@ -0,0 +1,207 @@
+
+	"github.com/matrixorigin/matrixone/pkg/container/ring"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/approxcd"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/avg"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/count"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/max"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/min"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/starcount"
+	"github.com/matrixorigin/matrixone/pkg/container/ring/sum"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+)
+
+var sumReturnTypes = map[types.T]types.T{
+	types.T_int8:    types.T_int64,
+	types.T_int16:   types.T_int64,
+	types.T_int32:   types.T_int64,
+	types.T_int64:   types.T_int64,
+	types.T_uint8:   types.T_uint64,
+	types.T_uint16:  types.T_uint64,
+	types.T_uint32:  types.T_uint64,
+	types.T_uint64:  types.T_uint64,
+	types.T_float32: types.T_float64,
+	types.T_float64: types.T_float64,
+}
+
+func ReturnType(op int, typ types.T) types.T {
+	switch op {
+	case Avg:
+		return types.T_float64
+	case Max:
+		return typ
+	case Min:
+		return typ
+	case Sum:
+		return sumReturnTypes[typ]
+	case Count, StarCount:
+		return types.T_int64
+	case ApproxCountDistinct:
+		return types.T_uint64
+	case Variance:
+		return types.T_float64
+	case BitAnd:
+		return types.T_uint64
+	case BitXor:
+		return types.T_uint64
+	case BitOr:
+		return types.T_uint64
+	case StdDevPop:
+		return types.T_float64
+	}
+	return 0
+}
+
+func New(op int, typ types.Type) (ring.Ring, error) {
+	switch op {
+	case Sum:
+		return NewSum(typ)
+	case Avg:
+		return avg.NewAvg(typ), nil
+	case Max:
+		return NewMax(typ)
+	case Min:
+		return NewMin(typ)
+	case Count:
+		return count.NewCount(typ), nil
+	case StarCount:
+		return starcount.NewCount(typ), nil
+	case ApproxCountDistinct:
+		return approxcd.NewApproxCountDistinct(typ), nil
+	case Variance:
+		return variance.NewVarianceRingWithTypeCheck(typ)
+	case BitAnd:
+		return NewBitAnd(typ)
+	case BitXor:
+		return NewBitXor(typ)
+	case BitOr:
+		return NewBitOr(typ)
+	case StdDevPop:
+		return stddevpop.NewStdDevPopRingWithTypeCheck(typ)
+	}
+	return nil, nil
+}
+
+func NewBitAnd(typ types.Type) (ring.Ring, error) {
+	switch typ.Oid {
+	case types.T_uint8, types.T_uint16, types.T_uint32, types.T_uint64, types.T_int8, types.T_int16, types.T_int32, types.T_int64, types.T_float32, types.T_float64:
+		return bitand.NewNumeric(typ), nil
+	}
+	return nil, errors.New(fmt.Sprintf("'%v' not support BitAnd", typ))
+}
+
+func NewBitOr(typ types.Type) (ring.Ring, error) {
+	switch typ.Oid {
+	case types.T_uint8, types.T_uint16, types.T_uint32, types.T_uint64, types.T_int8, types.T_int16, types.T_int32, types.T_int64, types.T_float32, types.T_float64:
+		return bitor.NewBitOr(typ), nil
+	}
+	return nil, errors.New(fmt.Sprintf("'%v' not support BitOr", typ))
+}
+
+func NewBitXor(typ types.Type) (ring.Ring, error) {
+	switch typ.Oid {
```

Comment:
Do you really need bitand/or/xor etc for float type?   What does that mean?

Reply:
I don't know. I'll have to ask the developer of this function about this one.